contributors/design-proposals/auditing.md

```diff
@@ -0,0 +1,340 @@
+    // AuditMetadata provides basic level of auditing, logging data at HTTP level
+    AuditMetadata AuditLevel   = "Metadata"
+    // AuditRequestBody provides Header level of auditing, and additionally
+    // logs unstructured request body
+    AuditRequestBody AuditLevel = "RequestBody"
+    // AuditResponseBody provides Request level of auditing, and additionally
+    // logs unstructured response body
+    AuditResponseBody AuditLevel = "ResponseBody"
+    // AuditStorageobject provides Response level, and additionally
+    // logs object before and after saving in storage
+    AuditStorageObject AuditLevel = "StorageObject"
+)
+```
+
+The audit level is determined by the policy, which maps a combination of requesting user, namespace, verb, API group, and resource. The policy is described in detail [below](#policy).
+
+When the http request is processed, the request handler will close the audit event by filling in the http response. Then it will pass the event to the configured output backend.
+
+**Note:** for service creation and deletion there is special REST code in the apiserver which takes care of service/node port (de)allocation and removal of endpoints on service deletion. Hence, these operations are not visible on the API layer and cannot be audit logged therefore. **No other resources** (with the exception of componentstatus which is not of interest here) **implement this kind of custom CRUD operations.**
+
+### Events
+
+The `Event` object contains the following data:
+
+```go
+package audit
+
+type Event struct {
+    // AuditLevel at which event was generated
+    Level AuditLevel
+
+    // below fields are filled at Metadata level and higher:
+
+    // Unique audit ID, generated for each request
+    ID string
+    // Time the event reached the apiserver
+    Timestamp Timestamp
+    // Source IP, from where the request originates
+    SourceIP string
+    // HTTP method sent by the client
+    HttpMethod string
+    // Authentication method used to allow users access the cluster
+    AuthMethod string
+    // RequestURI is the Request-Line as sent by the client to a server
+    RequestURI string
+    // User information
+    User UserInfo
+    // Impersonate-User header value
+    AsUser string
+    // Impersonate-Group header value
+    AsGroups string
+    // Object reference this request is targeted at
+    Object ObjectReference
+    // Response status code are returned by the server
+    ResponseStatusCode int
+    // Error response, if ResponseStatusCode >= 400
+    ResponseErrorMessage string
+
+    // below fields are filled at RequestObject level and higher:
+
+    // RequestObject logged before admission (json format)
+    RequestObject runtime.Unstructured
+    // Response object in json format
+    ResponseObject runtime.Unstructured
+
+    // below fields are filled at StorageObject level and higher:
+
+    // Object value before modification (will be empty when creating new object)
+    OldObject runtime.Object
```

Comment:
These are meant to be filled at the storage level, when an object is already converted and serialized, that's why it's structured objects.